Add note on PowerShell in Revision Selection

vegar opened the pull request:
<!-- Thanks for contributing! -->
<!-- Before you start on a large rewrite or other major change: open a new issue first, to discuss the proposed changes. -->
<!-- Should your changes appear in a printed edition, you'll be included in the contributors list. -->

<!-- Mark the checkbox [X] or [x] if you agree with the item. -->
- [x] I provide my work under the [project license](https://github.com/progit/progit2/blob/master/LICENSE.asc).
- [x] I grant such license of my work as is required for the purposes of future print editions to [Ben Straub](https://github.com/ben) and [Scott Chacon](https://github.com/schacon).

## Changes

- Add note on escaping curly braces in PowerShell when doing revision selection

## Context
<!--
List related issues.
Provide the necessary context to understand the changes you made.

Are you fixing a issue with this pull-request?
Use the "Fixes" keyword, to close the issue automatically after your work is merged.

Fixes #123
Fixes #456
-->
I've been struggling with the revision selection syntax `HEAD@{...}` and finally found out it's because of curly braces special meaning in PowerShell. 

I noticed the NOTE on special handling of caret in cmd.exe on Windows. A similar note about curly braces in PowerShell would have saved me some time and I hope to save fellow developers some time by adding in such a note.

HonkingGoose commented on book/07-git-tools/sections/revision-selection.asc:
```diff
@@ -180,6 +180,22 @@ Running `git show HEAD@{2.months.ago}` will show you the matching commit only if
+.Escaping brackets in PowerShell
+====
+
+On Windows in `powershell.exe`, `{` and `}` are special characters and needs to be treated differently.
```
```suggestion
When using PowerShell on Windows,  brackets like `{` and `}` are special characters and must be escaped.
```

HonkingGoose commented on book/07-git-tools/sections/revision-selection.asc:
```diff
@@ -180,6 +180,22 @@ Running `git show HEAD@{2.months.ago}` will show you the matching commit only if
+====
+
+On Windows in `powershell.exe`, `{` and `}` are special characters and needs to be treated differently.
+You can either escape them with a backtick ` or put the commit reference in quotes:
```
```suggestion
You can escape them with a backtick ` or put the commit reference in quotes:
```

HonkingGoose reviewed (commented):
Thanks for contributing! :+1: 

I have made some suggestions to improve the text. Let me know if you like/hate those. :smile: 

### Proposed changes:
- Rewrite first sentence to flow better, and to explain the bracket situation a bit better.
- Remove "either" from the second sentence, as I think the sentence is better without that word. The "or" in the middle of the sentence makes it clear that there are two choices to deal with the bracket issue. So "either" is not really adding value.

### How to merge these changes into your pull:

If you like my changes, you can put them in your pull request via the GitHub pull interface, read the
[GitHub docs, incorporating feedback in your pull request](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/incorporating-feedback-in-your-pull-request) for more on this.

HonkingGoose commented on book/07-git-tools/sections/revision-selection.asc:
```diff
@@ -180,6 +180,22 @@ Running `git show HEAD@{2.months.ago}` will show you the matching commit only if
+.Escaping brackets in PowerShell
+====
+
+When using PowerShell on Windows,  brackets like `{` and `}` are special characters and must be escaped.
```
```suggestion
When using PowerShell on Windows, brackets like `{` and `}` are special characters and must be escaped.
```

Fix whitespace, I made an error when making the suggestion. My bad! :disappointed:

vegar commented:
I’m totally fine having my non-native English improved. 😄 

One note: PowerShell has become cross-platform. I would guess the same issue is present in powershell on Linux, but I’m not sure. Maybe this should be checked out, or maybe this is good for now and removing the ‘on Windows’ part can be left for an improvement in the future.

HonkingGoose commented on book/07-git-tools/sections/revision-selection.asc:
```diff
@@ -180,6 +180,22 @@ Running `git show HEAD@{2.months.ago}` will show you the matching commit only if
+.Escaping brackets in PowerShell
+====
+
+When using PowerShell on Windows, brackets like `{` and `}` are special characters and must be escaped.
+You can escape them with a backtick ` or put the commit reference in quotes:
+
+[source,console]
+----
+$ git show HEAD@{0}     # will NOT work on Windows
+$ git show HEAD@`{0`}   # OK
+$ git show "HEAD@{0}"   # OK
+----
+
+====
+
```
```suggestion
When using PowerShell, brackets like `{` and `}` are special characters and must be escaped.
You can escape them with a backtick ` or put the commit reference in quotes:

[source,console]
----
$ git show HEAD@{0}     # will NOT work
$ git show HEAD@`{0`}   # OK
$ git show "HEAD@{0}"   # OK
----

====

```

I've just tested this on Linux. Using `git show HEAD@{0}` will not work on Linux either. It gives the following error:
`fatal: ambiguous argument 'HEAD@': unknown revision or path not in the working tree.`

Both "workarounds" work on Linux as well.

HonkingGoose commented on book/07-git-tools/sections/revision-selection.asc:
```diff
@@ -180,6 +180,22 @@ Running `git show HEAD@{2.months.ago}` will show you the matching commit only if
 ====
 
+[NOTE]
+.Escaping brackets in PowerShell
+====
+
+When using PowerShell, brackets like `{` and `}` are special characters and must be escaped.
+You can escape them with a backtick ` or put the commit reference in quotes:
+
+[source,console]
+----
+$ git show HEAD@{0}     # will NOT work
+$ git show HEAD@`{0`}   # OK
+$ git show "HEAD@{0}"   # OK
+----
+
+====
+
```
```suggestion
.Escaping braces in PowerShell
====

When using PowerShell, braces like `{` and `}` are special characters and must be escaped.
You can escape them with a backtick ` or put the commit reference in quotes:

[source,console]
----
$ git show HEAD@{0}     # will NOT work
$ git show HEAD@`{0`}   # OK
$ git show "HEAD@{0}"   # OK
----

====

```

I just learned a thing:
- Brackets: `[]`
- Braces: `{}`

Source: https://en.wikipedia.org/wiki/Bracket

vegar commented:
It ended up being quite a few minor change commits 😆 Do you know how it will be merged - if it gets merged? Will it be squashed?

HonkingGoose commented:
The maintainer usually does a normal merge, that would get all the minor change commits onto the mainline branch.
I've not really seen the maintainer do a squash-merge yet. :smile: 

I would recommend squashing the lot yourself. :smile: 

### How to squash the work into one nice clean commit:

These are the steps you can take on your local workstation. I'll assume you have Git installed on your workstation.

1. Ensure you have a clone of your own fork: `git clone https://github.com/vegar/progit2.git`
2. Go to your cloned directory. `cd directory_where_your_cloned_repo_is`
3. Add the original ProGit2 project as a upstream remote `git remote add upstream https://github.com/progit/progit2.git`
4. Fetch everything from your remote and from the upstream with `git fetch --all --prune`
5. Change to your `patch-1` branch. `git checkout patch-1` or `git switch patch-1`
6. Confirm you're on the right branch with `git status`.
7. Now enter the interactive rebase mode, by doing: `git rebase -i upstream/master`, your commit message editor will open with a list looking like this:

```
pick af3c02a Add note in Revision Selection about PowerShell 
pick 1388650 Update book/07-git-tools/sections/revision-selection.asc
pick 2876f1c Update book/07-git-tools/sections/revision-selection.asc
pick 4f94b8f Update book/07-git-tools/sections/revision-selection.asc
```

8. Squash the minor change commits into you first commit by selecting `s` for squash.

```
pick af3c02a Add note in Revision Selection about PowerShell 
s 1388650 Update book/07-git-tools/sections/revision-selection.asc
s 2876f1c Update book/07-git-tools/sections/revision-selection.asc
s 4f94b8f Update book/07-git-tools/sections/revision-selection.asc
```

9. Save the interactive rebase commit list, your commit editor will open a second file with a list of commit messages.
10. I would just have one commit message looking like this:

```
Add note in Revision Selection about Powershell

I've been struggling with the revision selection syntax `HEAD@{...}` and finally found out it's because of curly braces special meaning in PowerShell. 

I noticed the NOTE on special handling of caret in cmd.exe on Windows. A similar note about curly braces in PowerShell would have saved me some time and I hope to save fellow developers some time by adding in such a note.

Thanks to HonkingGoose for his minor changes.


Co-authored-by: HonkingGoose <34918129+HonkingGoose@users.noreply.github.com>
```

You do need the 2 newlines after the thank you message for the co-authored by section to work properly. So that's not a typo... :smile: 

11. `git push --force-with-lease` to push your nice commit up to the branch.

I hope this is helpful.

If you mess things up, you can use the `git rebase --abort` or get your work back from the `git reflog`.

You can also read https://git-rebase.io/ where you learn how to setup a sandbox environment to try this out before doing it for real.

Also maybe just copy the changed file and put it somewhere safe. That way you can always just restore from that file. :smile:

vegar commented:
Wow. That was quite some instructions. Hope :-D
I'm quite used to rebasing, but not the co-author ting. I read about it for the first time a couple of days ago, actually.
Anyway, without looking at your proposal, I would have kept my original commit and then one additional commit with all your improvements.  Wouldn't that be ok? Just to keep the history of the original vs the improved note?

I guess the comments in this pull request will go a little bananas after the force push, though...

vegar commented:
Had to push it twice to be sure I added the extra line feed. Looks like the double line feed doesn't show, but the co-author seems to work. 

Hope you're satisfied with the three commits. I kinda like to keep things separate, so I extracted the linux vs windows thingy into a separate commit just to make it clear that this is for both operating systems.

HonkingGoose commented:
>Had to push it twice to be sure I added the extra line feed. Looks like the double line feed doesn't show, but the co-author seems to work.

Yeah the double line feed does not really show up in the GitHub commit interface, but the co-author thing works. :+1: 

> Hope you're satisfied with the three commits. I kinda like to keep things separate, so I extracted the linux vs windows thingy into a separate commit just to make it clear that this is for both operating systems.

I'm happy with your 3 commits. I like your commit messages.

> Anyway, without looking at your proposal, I would have kept my original commit and then one additional commit with all your improvements. Wouldn't that be ok? Just to keep the history of the original vs the improved note?

For me, I like to squash everything into one commit, that single commit contains all changes. That way there is no chance of ending up with a invalid state when doing a `git revert` on part of the work but not on all of the work. 
But your way is clear and valid too. :+1:  If anybody wants to revert the work, they'll need to do the revert on the merge commit. :smile:

vegar commented:
Unless they only want to revet e.g. the linux vs windows part. :-)
I guess this is part of the reason why I like git so much: It's not just a tool to stash away old source code. It's a tool for communicating around code. If anyone reviews this change and thinks 'hey - powershell is a windows thing and the text should say so', they now have a commit telling that no, it's not a windows only thing. So even though the end result is the same it gives the opportunity to see a little bit of the thought process behind the end result as well.

HonkingGoose commented:
I'd say we're ready for a review by the head maintainer.

Hi @ben! :smile: :wave:  Can you take a look and tell us what you think?

ben reviewed (approved):
It's been a long journey, but this looks great. Thanks! ✨

TerriTauranga commented:
``
